lib/cfa/augeas_parser.rb

```diff
@@ -112,33 +142,25 @@ def save_to_augeas(aug, prefix)
       arrays = {}
 
       @data.each do |entry|
-        aug_key = obtain_aug_key(prefix, entry, arrays)
-        if entry[:value].is_a? AugeasTree
-          entry[:value].save_to_augeas(aug, aug_key)
-        else
-          report_error(aug) unless aug.set(aug_key, entry[:value])
-        end
+        save_entry(entry[:key], entry[:value], arrays, aug, prefix)
       end
     end
 
-    # @note for debugging purpose only
-    def dump_tree(prefix = "")
-      arrays = {}
+  private
 
-      @data.each_with_object("") do |entry, res|
-        aug_key = obtain_aug_key(prefix, entry, arrays)
-        if entry[:value].is_a? AugeasTree
-          res << entry[:value].dump_tree(aug_key)
-        else
-          res << aug_key << "\n"
-        end
+    def save_entry(key, value, arrays, aug, prefix)
+      aug_key = obtain_aug_key(prefix, key, arrays)
+      case value
+      when AugeasTree then value.save_to_augeas(aug, aug_key)
+      when AugeasTreeValue
+        report_error(aug) unless aug.set(aug_key, value.value)
+        value.tree.save_to_augeas(aug, aug_key)
+      else
+        report_error(aug) unless aug.set(aug_key, value)
       end
     end
 
-  private
-
-    def obtain_aug_key(prefix, entry, arrays)
-      key = entry[:key]
+    def obtain_aug_key(prefix, key, arrays)
       if key.end_with?("[]")
         array_key = key.sub(/\[\]$/, "")
```

Comment:
NP: This looks too complicated, regexp (with escapes) substitution with empty string....

What about simple `key.chomp("[]")` ? (See http://ruby-doc.org/core-2.3.1/String.html#method-i-chomp)